cloudflare/resource_cloudflare_waf_package_test.go

```diff
@@ -0,0 +1,117 @@
+package cloudflare
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	cloudflare "github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccCloudflareWAFPackage_CreateThenUpdate(t *testing.T) {
+	t.Parallel()
+	zoneID := os.Getenv("CLOUDFLARE_ZONE_ID")
+	packageID, err := testAccGetWAFPackage(zoneID)
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+
+	rnd := generateRandomResourceName()
+	name := fmt.Sprintf("cloudflare_waf_package.%s", rnd)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCloudflareWAFPackageDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflareWAFPackageConfig(zoneID, packageID, "medium", "simulate", rnd),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(name, "package_id", packageID),
+					resource.TestCheckResourceAttr(name, "zone_id", zoneID),
+					resource.TestCheckResourceAttr(name, "sensitivity", "medium"),
+					resource.TestCheckResourceAttr(name, "action_mode", "simulate"),
+				),
+			},
+			{
+				Config: testAccCheckCloudflareWAFPackageConfig(zoneID, packageID, "low", "block", rnd),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(name, "package_id", packageID),
+					resource.TestCheckResourceAttr(name, "zone_id", zoneID),
+					resource.TestCheckResourceAttr(name, "sensitivity", "low"),
+					resource.TestCheckResourceAttr(name, "action_mode", "block"),
+				),
+			},
+		},
+	})
+}
+
+func testAccGetWAFPackage(zoneID string) (string, error) {
+	config := Config{}
+	if apiToken, ok := os.LookupEnv("CLOUDFLARE_API_TOKEN"); ok {
```

Comment:
I don't really like this in here as we're duplicating the logic we already have in the provider setup. If that ever changes, we now have somewhere that depends on it directly. Was there a reason you couldn't use the [`sharedClient`](https://github.com/terraform-providers/terraform-provider-cloudflare/blob/7df730622f105a6737afedd2c24453c855b32322/cloudflare/cloudflare_sweeper_test.go#L17)? We're already using that in sweepers and it leverages the existing methods.

Reply:
If we can't use the `shareClient` functionality, let's just copy the function and use it here. Don't worry about the logic as we can swap it all when/if we move to API tokens for the test suite.

Another drawback to having this logic here is that `testAccPreCheck` doesn't check for the API token, only the older method of authenticating.

Reply:
> Another drawback to having this logic here is that testAccPreCheck doesn't check for the API token, only the older method of authenticating which means use of API tokens isn't enforced at all.

Note that `testAccPreCheck` is still run just after, so any check it does will be done here, so that should be fine ? But true that we're listing the packages before having done that check...

> I don't really like this in here as we're duplicating the logic we already have in the provider setup. If that ever changes, we now have somewhere that depends on it directly. Was there a reason you couldn't use the sharedClient? We're already using that in sweepers and it leverages the existing methods.

I get this when using the `sharedClient`:

```
    resource_cloudflare_waf_package_test.go:18: Error while listing WAF packages: error from makeRequest: HTTP status 400: content "{\"success\":false,\"errors\":[{\"code\":6003,\"message\":\"Invalid request headers\",\"error_chain\":[{\"code\":6103,\"message\":\"Invalid format for X-Auth-Key header\"}]}],\"messages\":[],\"result\":null}"
```

Which means the `client.ListWAFPackages` is failing. Any idea what would be missing ?

Reply:
> `sharedClient` currently seems to be forcing to use API Key & email instead of API token, is this expected ?

Yep, we'll eventually move this over to API tokens when it's supported fully.

WRT to your error, are you setting the `CLOUDFLARE_API_KEY` and `CLOUDFLARE_EMAIL` per the client initialisation when you're running the tests? As you mentioned, the PreCheck usually handles this but because we're running before it, it could be missed.

Reply:
I did not want to go fetch an API key so I used `CLOUDFLARE_API_KEY="blah"` for my tests above, which is why I had that error (this is also why I corrected my message after the fact, realizing the error was related to that).

I think my last push (fixup this time, sorry again) should be good to go ? (using `sharedClient`)

Reply:
Gotcha, that makes total sense then 😄